implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheusmetrics/PrometheusNamingConvention.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2017 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.prometheusmetrics;
+
+import io.micrometer.common.lang.Nullable;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.config.NamingConvention;
+
+import java.util.regex.Pattern;
+
+/**
+ * See https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels for a
+ * specification of the constraints on metric names and labels
+ *
+ * @author Jon Schneider
+ */
+public class PrometheusNamingConvention implements NamingConvention {
+
+    private static final Pattern nameChars = Pattern.compile("[^a-zA-Z0-9_:]");
+
+    private static final Pattern tagKeyChars = Pattern.compile("[^a-zA-Z0-9_]");
+
+    private final String timerSuffix;
+
+    public PrometheusNamingConvention() {
+        this("");
+    }
+
+    public PrometheusNamingConvention(String timerSuffix) {
+        this.timerSuffix = timerSuffix;
+    }
+
+    /**
+     * Names are snake-cased. They contain a base unit suffix when applicable.
+     * <p>
+     * Names may contain ASCII letters and digits, as well as underscores and colons. They
+     * must match the regex [a-zA-Z_:][a-zA-Z0-9_:]*
+     */
+    @Override
+    public String name(String name, Meter.Type type, @Nullable String baseUnit) {
```

Comment:
I think we can delegate to `PrometheusNaming` in the implementation of our default `NamingConvention`. The important point was for the configured `NamingConvention` to be used instead of hard-coding using `PrometheusNaming` in `PrometheusMeterRegistry` and ignoring what the user configured.

Reply:
We can utilize `PrometheusNaming` but we still need to keep our custom logic in the naming convention where we append units and other things.